addons/mass_mailing/models/mail_notification.py

```diff
@@ -49,53 +40,59 @@ class MailingTrace(models.Model):
     # Status
-    state = fields.Selection(compute="_compute_state",
-                             selection=[('outgoing', 'Outgoing'),
-                                        ('sent', 'Sent'),
-                                        ('opened', 'Opened'),
-                                        ('replied', 'Replied'),
-                                        ('bounced', 'Bounced'),
-                                        ('exception', 'Exception'),
-                                        ('canceled', 'Canceled')], store=True)
-    failure_type = fields.Selection(selection=[
-        ("SMTP", "Connection failed (outgoing mail server problem)"),
-        ("RECIPIENT", "Invalid email address"),
-        ("BOUNCE", "Email address rejected by destination"),
-        ("UNKNOWN", "Unknown error")], string='Failure type')
-    state_update = fields.Datetime(compute="_compute_state", string='State Update',
-                                   help='Last state update of the mail',
-                                   store=True)
-
-    @api.depends('trace_type', 'mass_mailing_id')
+    notification_status = fields.Selection(selection_add=[
+        ('outgoing', 'Outgoing'),
+        ('opened', 'Opened'),
+        ('replied', 'Replied')])
+    status_update = fields.Datetime(string='State Update', help='Last state update of the mail', store=True)
+
+    @api.depends('mass_mailing_id.campaign_id')
+    def _compute_campaign_id(self):
+        for notif in self:
+            notif.campaign_id = notif.mass_mailing_id.campaign_id
+
+
+    @api.depends('notification_type', 'mass_mailing_id', 'res_partner_id')
     def _compute_display_name(self):
         for trace in self:
-            trace.display_name = '%s: %s (%s)' % (trace.trace_type, trace.mass_mailing_id.name, trace.id)
-
-    @api.depends('sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled')
-    def _compute_state(self):
-        self.update({'state_update': fields.Datetime.now()})
-        for stat in self:
-            if stat.canceled:
-                stat.state = 'canceled'
-            elif stat.exception:
-                stat.state = 'exception'
-            elif stat.opened or stat.clicked:
-                stat.state = 'opened'
-            elif stat.replied:
-                stat.state = 'replied'
-            elif stat.bounced:
-                stat.state = 'bounced'
-            elif stat.sent:
-                stat.state = 'sent'
-            else:
-                stat.state = 'outgoing'
+            trace.display_name = '%s: %s (%s)' % (trace.notification_type,
+                                                  trace.mass_mailing_id.name or trace.res_partner_id.display_name or '',
+                                                  trace.id)
+
+    def _update_trace_status(self, values):
+        if set(['sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled']) & set(values.keys()):
```

Comment:
Oh boy.

`if any(values.keys() in ['sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled'])` ?

Reply:
It can't be work, generator in list/set doesn't work. but I simplify the first version by 
values.keys() & {'sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled'}:

Reply:
Then maybe

`if any(list(values.keys()) in ['sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled']) ?`

?

Reply:
Nop, it is not the same, I need the intersection (list in list doesn't work  in python)

Reply:
I'm sorry, I meant:

`if any(key in ['sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled'] for key in values.keys())`

Reply:
If you want, I didn't see the benefit again the other solution

Reply:
It's a matter of preferences.

I think it's easier to read:
"If any of my keys is contained in my whitelist"
then:
"If the intersection of my keys and my whitelist is not empty"

I won't blame you if you want to keep your version :)